common/flatpak-dir.c

```diff
@@ -8223,27 +8309,85 @@ flatpak_dir_find_remote_ref (FlatpakDir   *self,
-                                      opt_default_branch, opt_arch, collection_id,
-                                      kinds, out_kind, &my_error);
-  if (!remote_ref)
+  if (collection_id == NULL)
     {
-      if (g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+      if (!flatpak_dir_remote_list_refs (self, remote,
+                                         &remote_refs, cancellable, error))
+        return NULL;
+
+      remote_ref = find_ref_for_refs_set (remote_refs, name, opt_branch,
+                                          opt_default_branch, opt_arch, NULL, kinds,
+                                          out_kind, &my_error);
+
+      if (!remote_ref)
         {
-          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
-                       _("Error searching remote %s: %s"),
-                       remote,
-                       my_error->message);
-          return NULL;
+          if (g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                           _("Error searching remote %s: %s"),
+                           remote,
+                           my_error->message);
+              return NULL;
+            }
+          else
+            {
+              g_propagate_error (error, g_steal_pointer (&my_error));
+              return NULL;
+            }
+        }
+    }
+  else
+    {
+      g_autoptr(GVariant) xa_cache = NULL;
+      g_autoptr(GPtrArray) matched_refs = NULL;
+      gsize j;
+
+      /* Use xa.cache because we can't fetch the summary when offline */
+      if (!flatpak_dir_fetch_refs_cache (self, remote, &xa_cache, cancellable, error))
+        return NULL;
+
+      matched_refs = find_matching_refs_from_cache (xa_cache,
+                                                    collection_id,
+                                                    name,
+                                                    opt_branch,
+                                                    opt_arch,
+                                                    kinds,
+                                                    FIND_MATCHING_REFS_FLAGS_NONE,
+                                                    error);
+      if (matched_refs == NULL)
+        return NULL;
+      else if (matched_refs->len == 1)
+        {
+          remote_ref = g_strdup (g_ptr_array_index (matched_refs, 0));
+          if (out_kind != NULL)
+            {
+              if (g_str_has_prefix (remote_ref, "app/"))
+                *out_kind = FLATPAK_KINDS_APP;
+              else
+                *out_kind = FLATPAK_KINDS_RUNTIME;
+            }
         }
       else
         {
-          g_propagate_error (error, g_steal_pointer (&my_error));
+          g_autoptr(GString) err = g_string_new ("");
+          g_string_printf (err, _("Multiple branches available for %s, you must specify one of: "), name);
```

Comment:
The code at https://github.com/endlessm/flatpak/blob/47536627212bd3f60647ce49f943ec2e7fcba890/common/flatpak-dir.c#L8124 looks at `opt_default_branch` to determine if one of these is the one we're looking for. Perhaps we should do the same here, and factor that code out into a helper function.